Simplify and extend ValueRanges

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102611
* #104558
* #104559
* __->__ #104557

This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8

lezcano committed:
Simplify and extend ValueRanges

This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

[ghstack-poisoned]

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -222,25 +212,23 @@ def eq(a, b):
-            return ValueRanges.wrap(sympy.false)
-        return ValueRanges(sympy.false, sympy.true)
+        assert a.is_bool == b.is_bool
+        if a.is_bool:
+            return cls.and_(cls.not_(a), b)
```
I find the following easier to follow than  `cls.and_(cls.not_(a), b)`:
```
if a.is_bool:
     a, b = (ValueRanges.increasing_map(val, lambda x: bool(x)) for val in (a, b))

if a.upper < b.lower:
...
```

Also I think `cls.and_(cls.not_(a), b)` gives wrong results for :


`False True < False True`.   `(True False) and (False True) -> False False` but we should be getting `False True`

lezcano replied:
I don't think I follow your reasoning with the increasing map thing.

The thing here is that `a < b` for `a, b` bool is mathematically defined as `~a & b`. That's what we're using here.
Your example is wrong, Note that `not([False, True]) = [False, True]`. Your range is flipped. In fact, we check in the ValueRanges constructor that we never get ranges for which `lower > upper`.

eellison replied:
Shouldn't we be following python semantics here ? False < True -> True.

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -252,37 +240,54 @@ def ge(cls, a, b):
-        if a.is_singleton():
-            return ValueRanges.wrap(sympy.Not(a.lower))
-        return ValueRanges(sympy.false, sympy.true)
+        return ValueRanges.decreasing_map(a, sympy.Not)
```
nit: maybe keep the assert is_bool

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -323,46 +328,48 @@ def floordiv(a, b):
-                return result
-            return ValueRanges.coordinatewise_monotone_map(a, b, f)
+    @classmethod
+    def truncdiv(cls, a, b):
+        x = cls.truediv(a, b)
+        if x == ValueRanges.unknown():
+            return x
```
is this equivalent, or fixing a bug ?

lezcano replied:
Equivalent and noticeably simpler.

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -323,46 +328,48 @@ def floordiv(a, b):
+            # Make unknown() * wrap(0) == wrap(0)
             if a == 0:
-                return 0
+                return a
```
jc, why did you have to return a here instead of 0?

lezcano replied:
to return a zero that's both a sympy object and it's of the correct dtype.

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -401,28 +412,46 @@ def is_integer(val):
+        if b < 0:
+            a = cls.reciprocal(a)
+            b = -b
+        acc = ValueRanges.wrap(1)
```
I wonder if this is a bit slow for what it buys us, but I guess it's fine

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -401,28 +412,46 @@ def is_integer(val):
+            a = cls.reciprocal(a)
+            b = -b
+        acc = ValueRanges.wrap(1)
+        while b > 0:
+            if b % 2 == 1:
+                acc = cls.mul(acc, a)
+            a = cls.mul(a, a)
+            b = b // 2
```
Can we not just call into sympy.pow at this point ? Feel like there's probably a way of implementing this simpler

lezcano replied:
Yeah, let me rewrite this. This one's a weird op, as `-3.0 ** x` is not well defined, but `-3 ** x` is well defined for x integer. I'll split it into `a` integer and not.

lezcano replied:
Ok, I heavily simplified this logic, now it looks much better.

eellison reviewed (commented):
Nice !! Have a few comments. 

Is it worth committing the fuzzing script that Nuno wrote https://gist.github.com/nunoplopes/dc08bee4f3fb3609f916aa1e9389273a ? Or something like it, maybe as a slow test.

lezcano commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -54,16 +54,17 @@ def sympy_generic_le(lower, upper):
+    # with constant sympy expressions
     lower: Union[sympy.Expr, SympyBoolean]
     upper: Union[sympy.Expr, SympyBoolean]
+    is_bool: bool
```
Yeah. It's in my radar to improve it eventually to track the type properly and make sure that the type of the bounds and the type of the object is sound. For now, bool it is, given that booleans are just very different to floats and integers in sympy.

lezcano commented:
For the testing, I plan to push some proper testing soonish, where we test these ranges on the FX interpreter, for the nodes that we use in that interpreter, and then keep Ed's tests for the SymPy nodes. I also plan to trim the SymPy interpreter to just implement the functions we currently use for the guards and the indexing on inductor.

lezcano commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -19,7 +21,9 @@ class ValueRangeError(RuntimeError):
 # sympy expressions don't have free variables
 def simple_sympify(e):
-    if isinstance(e, int):
+    if isinstance(e, bool):
+        return sympy.true if e else sympy.false
```
good catch

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

lezcano commented:
@eellison addressed the review

lezcano commented:
The next two are also ready and should be quite easy reviews :)

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78

[ghstack-poisoned]

lezcano committed:
Update on "Simplify and extend ValueRanges"


This PR:
- It adds a few boolean variants of some methods that were missing
- It simplifies the implementation of plenty of the operations
- Adds ModularIndexing to the SymPy interpreter

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]
